File: RealtimeTranslator/Tests/SwitchStabilityFixtureTests.swift
import XCTest
@testable import RealtimeTranslator

final class SwitchStabilityFixtureTests: XCTestCase {
    // Given: 同じ原文を異なる delta 分割で表した switchStability fixture
    // When: processor と同じ routing window / tracker / selector の流れを各分割へ適用する
    // Then: 切替回数・最終 target・最初の split offset が分割に依存しない
    func testSwitchStabilityMatchesSharedFixture() throws {
        for fixture in try SharedFixtures.section("routing", "switchStability") {
            let pair = try XCTUnwrap(
                LanguagePair(rawValue: SharedFixtures.text(fixture["pair"]))
            )
            let initialTarget = try XCTUnwrap(
                RealtimeTranslationOutputLanguage(
                    rawValue: SharedFixtures.text(fixture["initialTarget"])
                )
            )
            let splits = try XCTUnwrap(fixture["splits"] as? [[String: Any]])
            var joinedText: String?
            var firstSplitOffset: Int?

            for split in splits {
                let deltas = try XCTUnwrap(split["deltas"] as? [String])
                let splitText = deltas.joined()
                if let joinedText {
                    XCTAssertEqual(splitText, joinedText, SharedFixtures.text(fixture["name"]))
                } else {
                    joinedText = splitText
                }

                var target = initialTarget
                var source = ""
                var routing = ""
                var tracker = SourceBoundaryTracker()
                var reverseEvidenceCount = 0
                var switchCount = 0
                var switchDelta: Int?
                var sourceLengthAtSwitch: Int?
                var splitOffset: Int?

                for (index, delta) in deltas.enumerated() {
                    let deltaStart = source.utf16.count
                    source += delta
                    routing = RoutingSourceTextWindow.trim(routing + delta, pair: pair)
                    let evidence = SpokenLanguageDetector.recentEvidence(
                        in: routing,
                        pair: pair
                    )
                    let currentLanguage = try XCTUnwrap(pair.counterpart(of: target))
                    tracker.observe(
                        segmentSource: source,
                        deltaStart: deltaStart,
                        segmentGeneration: 0,
                        pair: pair,
                        currentLanguage: currentLanguage,
                        reverseEvidenceCount: 0
                    )
                    let selection = TranslationTargetSelector.select(
                        pair: pair,
                        currentTarget: target,
                        reverseEvidenceCount: reverseEvidenceCount,
                        evidence: evidence,
                        oppositeRun: tracker.oppositeScriptRun(in: source)
                    )
                    reverseEvidenceCount = selection.reverseEvidenceCount

                    guard let nextTarget = selection.target, nextTarget != target else {
                        continue
                    }

                    let offset = tracker.candidateOffset ?? deltaStart
                    if switchCount == 0 {
                        switchDelta = index
                        sourceLengthAtSwitch = source.utf16.count
                        splitOffset = offset
                    }
                    switchCount += 1
                    let splitIndex = String.Index(utf16Offset: offset, in: source)
                    source = String(source[splitIndex...])
                    routing = RoutingSourceTextWindow.trim(source, pair: pair)
                    tracker.reset()
                    target = nextTarget
                    reverseEvidenceCount = 0
                }

                XCTAssertEqual(
                    switchCount,
                    SharedFixtures.number(fixture["expectedSwitchCount"]),
                    SharedFixtures.text(fixture["name"])
                )
                XCTAssertEqual(
                    target.rawValue,
                    SharedFixtures.text(fixture["expectedFinalTarget"]),
                    SharedFixtures.text(fixture["name"])
                )
                XCTAssertEqual(
                    switchDelta,
                    SharedFixtures.optionalNumber(split["expectedSwitchAtDelta"]),
                    SharedFixtures.text(fixture["name"])
                )
                XCTAssertEqual(
                    sourceLengthAtSwitch,
                    SharedFixtures.optionalNumber(split["expectedSourceLengthAtSwitch"]),
                    SharedFixtures.text(fixture["name"])
                )
                if firstSplitOffset == nil {
                    firstSplitOffset = splitOffset
                } else {
                    XCTAssertEqual(splitOffset, firstSplitOffset, SharedFixtures.text(fixture["name"]))
                }
            }

            XCTAssertEqual(
                firstSplitOffset,
                SharedFixtures.optionalNumber(fixture["expectedSplitOffset"]),
                SharedFixtures.text(fixture["name"])
            )
        }
    }
}

File: windows/tests/RealtimeTranslator.Core.Tests/SwitchStabilityFixtureTests.cs
using System;
using System.Linq;
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.Realtime;
using RealtimeTranslator.Core.OpenAI;
using Xunit;

namespace RealtimeTranslator.Core.Tests;

public sealed class SwitchStabilityFixtureTests
{
    public static TheoryData<string> Cases =>
        SharedFixtures.CaseNames("routing", "switchStability");

    // Given: 同じ原文を異なる delta 分割で表した switchStability fixture
    // When: processor と同じ routing window / tracker / selector の流れを各分割へ適用する
    // Then: 切替回数・最終 target・最初の split offset が分割に依存しない
    [Theory]
    [MemberData(nameof(Cases))]
    public void SwitchStabilityMatchesFixture(string name)
    {
        var fixture = SharedFixtures.Case("routing", "switchStability", name);
        var pair = LanguagePairExtensions.ParseLanguagePair(
            SharedFixtures.Text(fixture["pair"]));
        var initialTarget = RealtimeTranslationWireValues.ParseOutputLanguage(
            SharedFixtures.Text(fixture["initialTarget"]));
        string? joinedText = null;
        int? firstSplitOffset = null;

        foreach (var splitNode in fixture["splits"]!.AsArray())
        {
            var split = splitNode!.AsObject();
            var deltas = split["deltas"]!.AsArray();
            var splitText = string.Concat(deltas.Select(SharedFixtures.Text));
            if (joinedText is null)
            {
                joinedText = splitText;
            }
            else
            {
                Assert.Equal(joinedText, splitText);
            }

            var target = initialTarget;
            var source = string.Empty;
            var routing = string.Empty;
            var tracker = new SourceBoundaryTracker();
            var reverseEvidenceCount = 0;
            var switchCount = 0;
            int? switchDelta = null;
            int? sourceLengthAtSwitch = null;
            int? splitOffset = null;

            for (var index = 0; index < deltas.Count; index += 1)
            {
                var delta = SharedFixtures.Text(deltas[index]);
                var deltaStart = source.Length;
                source += delta;
                routing = RoutingSourceTextWindow.Trim(routing + delta, pair);
                var evidence = SpokenLanguageDetector.RecentEvidence(routing, pair);
                var currentLanguage = pair.Counterpart(target)
                    ?? throw new Xunit.Sdk.XunitException("missing current language");
                tracker.Observe(
                    source,
                    deltaStart,
                    0,
                    pair,
                    currentLanguage,
                    0);
                var selection = TranslationTargetSelector.Select(
                    pair,
                    target,
                    reverseEvidenceCount,
                    evidence,
                    tracker.OppositeScriptRun(source));
                reverseEvidenceCount = selection.ReverseEvidenceCount;

                if (selection.Target is not { } nextTarget || nextTarget == target)
                {
                    continue;
                }

                var offset = tracker.CandidateOffset ?? deltaStart;
                if (switchCount == 0)
                {
                    switchDelta = index;
                    sourceLengthAtSwitch = source.Length;
                    splitOffset = offset;
                }

                switchCount += 1;
                source = source[offset..];
                routing = RoutingSourceTextWindow.Trim(source, pair);
                tracker.Reset();
                target = nextTarget;
                reverseEvidenceCount = 0;
            }

            Assert.Equal(
                SharedFixtures.Number(fixture["expectedSwitchCount"]),
                switchCount);
            Assert.Equal(
                SharedFixtures.Text(fixture["expectedFinalTarget"]),
                target.ToWireValue());
            Assert.Equal(
                SharedFixtures.OptionalNumber(split["expectedSwitchAtDelta"]),
                switchDelta);
            Assert.Equal(
                SharedFixtures.OptionalNumber(split["expectedSourceLengthAtSwitch"]),
                sourceLengthAtSwitch);
            if (firstSplitOffset is null)
            {
                firstSplitOffset = splitOffset;
            }
            else
            {
                Assert.Equal(firstSplitOffset, splitOffset);
            }
        }

        Assert.Equal(
            SharedFixtures.OptionalNumber(fixture["expectedSplitOffset"]),
            firstSplitOffset);
    }
}
